api/handlers.go

```diff
@@ -0,0 +1,109 @@
+package api
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/ONSdigital/dp-recipe-api/recipe"
+	"github.com/ONSdigital/log.go/log"
+	"github.com/gorilla/mux"
+)
+
+//RecipeListHandler - get all recipes
+// USAGE: curl -X GET http://localhost:22300/recipes
+func (api *RecipeAPI) RecipeListHandler(w http.ResponseWriter, req *http.Request) {
+	var list recipe.List
+	if api.EnableMongoData {
+		var err error
+		list.Items, err = api.dataStore.Backend.GetRecipes()
+		if err != nil {
+			log.Event(req.Context(), "error getting recipes from mongo", log.ERROR, log.Error(err))
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	} else {
+		list = recipe.FullList
+	}
+
+	c := len(list.Items)
+	list.Count = c
+	list.TotalCount = c
+	list.Limit = c
+
+	b, err := json.Marshal(list)
+	if err != nil {
+		log.Event(req.Context(), "error returned from json marshall", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(b)
+}
+
+//RecipeHandler - get recipe by ID
+// USAGE: curl -X GET http://localhost:22300/recipes/{id}
+func (api *RecipeAPI) RecipeHandler(w http.ResponseWriter, req *http.Request) {
+	vars := mux.Vars(req)
+	recipeID := vars["id"]
+	logD := log.Data{"recipe_id": recipeID}
+
+	var r recipe.Response
+	if api.EnableMongoData {
+
+		recipe, err := api.dataStore.Backend.GetRecipe(recipeID)
+		if err != nil {
+			log.Event(req.Context(), "recipe not found", log.ERROR, logD)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		r = *recipe
+
+	} else {
+
+		found := false
+
+		for _, item := range recipe.FullList.Items {
+			if item.ID == recipeID {
+				r = item
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			log.Event(req.Context(), "recipe not found", log.ERROR, logD)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+
+	b, err := json.Marshal(r)
+	if err != nil {
+		log.Event(req.Context(), "error returned from json marshall", log.ERROR, log.Error(err), logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(b)
+}
+
+//HealthCheck - health check endpoint
+func (api *RecipeAPI) HealthCheck(w http.ResponseWriter, req *http.Request) {
+	// Set status to 200 OK
+	w.WriteHeader(200)
+}
+
+//AddAllRecipeHandler - Adds all the recipes from data.go to the mongo database
+//USAGE: curl -X POST http://localhost:22300/allrecipes
+func (api *RecipeAPI) AddAllRecipeHandler(w http.ResponseWriter, req *http.Request) {
+	for _, item := range recipe.FullList.Items {
+		err := api.dataStore.Backend.AddRecipe(item)
+		if err != nil {
+			log.Event(req.Context(), "error in adding all recipes to mongo from data.go", log.ERROR, log.Error(err))
+			w.WriteHeader(http.StatusInternalServerError)
+			return
```

Comment:
maybe instead of returning here, we can complete the loop, keep track of any error, and if there is any error, write the InternalServerError at the end?
Something like:
```
hasErrors := false
for _, item := range recipe.FullList.Items {
    ...
    if err != nil {
        hasErrors = true
    }
...

if HasErrors {
    w.WriteHeader(http.StatusInternalServerError)
}
```


Another option (with better performance) would be to have a single query against mongodb to add all the items in a single query, if it is possible at all.

Reply:
But wouldn't that be inefficient? If an error is found in one of the recipe items, then it would continue to iterate over the remaining items the recipe list even if hasErrors has been set to true 

On the other hand, the return statement within the for loop would exit the loop when an error is found. 

Am I right in this?

Reply:
Yes, absolutely right! - but if we did the whole loop we would update as many items as possible, and just ignore the ones that we can't be updated - I don't know what would be the best option, I will accept it as it is now.